### PR TITLE
[Flink AI Flow][Plugins] Introduce WrappedTableEnvironment and WrappedStatementSet in Flink job plugin

### PR DESCRIPTION
**Pull request by lisy09**
## What is the purpose of the change

To fix the bug related to flink_run_main.py & FlinkPythonProcessor mentioned in https://github.com/alibaba/flink-ai-extended/issues/475

## Brief change log

  - add ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py
     -  introduce WrappedBatchTableEnvironment, WrappedStreamTableEnvironment, WrappedStatementSet, which
     - use wrapped class to **collect execution information when calling add_insert_sql and/or execute_sql**
  - replace original class by wrapped class in flink_run_main.py & flink_env.py
  - will write multiple job ids to ./job_id instead of single job id

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

(WIP: adding a branch to build flink&hive environment and test workflow which caused bug)

Integration test with flink & hive cluster in an external repo & branch:
https://github.com/lisy09/flink-ai-extended/tree/bug_fix/ai_flow/issue_475_test_case/flink-ai-flow/ai_flow_plugins/tests/issue_475

### CODE REVIEW

**Comment by lisy09** [comment]
@SteNicholas
Please help to check whether this approach works.
And I am now creating a another fork branch to build the test environment (flink & hive cluster).
Will back and report soon.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py [hunk 1]
```diff
@@ -29,7 +35,7 @@ class FlinkEnv(object):
     """
 
     @abstractmethod
-    def create_env(self) -> (ExecutionEnvironment, TableEnvironment, StatementSet):
+    def create_env(self) -> Tuple[ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet]:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py:38]:
```suggestion
    def create_env(self) -> (ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet):
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py [hunk 2]
```diff
@@ -38,11 +44,12 @@ class FlinkBatchEnv(FlinkEnv):
     FlinkBatchEnv is the default implementation of FlinkEnv, used in flink batch jobs.
     """
 
-    def create_env(self) -> (ExecutionEnvironment, TableEnvironment, StatementSet):
+    def create_env(self) -> Tuple[ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet]:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py:47]:
```suggestion
    def create_env(self) -> (ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet):
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py [hunk 3]
```diff
@@ -53,11 +60,12 @@ class FlinkStreamEnv(FlinkEnv):
     FlinkStreamEnv is the default implementation of FlinkEnv, used in flink streaming jobs.
     """
 
-    def create_env(self) -> (ExecutionEnvironment, TableEnvironment, StatementSet):
+    def create_env(self) -> Tuple[ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet]:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py:63]: Ditto.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            if job_client is not None:
+                job_id = str(job_client.get_job_id())
+                result.append(job_id)
+        return result
+
+    def wait_execution_results(self):
+        for r in self.execute_sql_results:
+            job_client = r.get_job_client()
+            if job_client is not None:
+                job_client.get_job_execution_result(
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py]: The `get_job_execution_result` has no parameter named the `user_class_loader`.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            if job_client is not None:
+                job_id = str(job_client.get_job_id())
+                result.append(job_id)
+        return result
+
+    def wait_execution_results(self):
+        for r in self.execute_results:
+            job_client = r.get_job_client()
+            if job_client is not None:
+                job_client.get_job_execution_result(
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py]: Ditto.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import List, Text
+from pyflink.table import StatementSet, TableEnvironment
+from pyflink.table.table_environment import BatchTableEnvironment, StreamTableEnvironment
+from pyflink.table.table_result import TableResult
+
+
+class WrappedTableEnvironmentContext:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:23]: Please add the comments for the class `WrappedTableEnvironmentContext`.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

**Comment by SteNicholas** [comment]
@lisy09 , please add the unit tests for the changes of PR.

**Comment by lisy09** [comment]
> @lisy09 , please add the unit tests for the changes of PR.

@SteNicholas Thanks for your review.
Will update later and back to you.

**Comment by lisy09** [comment]
@SteNicholas @wuchaochen 
I finished the change regarding last review, and add two test cases in ai_flow_plugins/tests/job_plugins/ut_workflows/workflows/test_flink/test_flink.py.

Please help to review the latest change .

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py [hunk 7]
```diff
@@ -29,7 +35,7 @@ class FlinkEnv(object):
     """
 
     @abstractmethod
-    def create_env(self) -> (ExecutionEnvironment, TableEnvironment, StatementSet):
+    def create_env(self) -> (ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet):
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py:38]: From an API perspective, the `create_env` interface should not contain the `WrappedTableEnvironment`,  `WrappedStatementSet`, which could move to the `flink_wrapped_env.py`.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py [hunk 8]
```diff
@@ -53,11 +60,12 @@ class FlinkStreamEnv(FlinkEnv):
     FlinkStreamEnv is the default implementation of FlinkEnv, used in flink streaming jobs.
     """
 
-    def create_env(self) -> (ExecutionEnvironment, TableEnvironment, StatementSet):
+    def create_env(self) -> (ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet):
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py:63]: Ditto.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py (hunk 7 above)
Comment (lisy09) [reply flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py]: fix with https://github.com/alibaba/flink-ai-extended/pull/477/commits/2eb4dea40f4d9841bb7cc8117be46ba5c08911e6

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py (hunk 8 above)
Comment (lisy09) [reply flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_env.py]: fix with https://github.com/alibaba/flink-ai-extended/pull/477/commits/2eb4dea40f4d9841bb7cc8117be46ba5c08911e6

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_run_main.py [hunk 9]
```diff
@@ -72,11 +72,17 @@ def close():
...
+        statement_set.execute()
+    
+    job_id_list = table_env.wrapped_context.get_job_ids() + statement_set.wrapped_context.get_job_ids()
+    if len(job_id_list) > 0:
         with open('./job_id', 'w') as fp:
-            fp.write(str(job_client.get_job_id()))
-        job_client.get_job_execution_result(user_class_loader=None).result()
+            fp.write(str(job_id_list))
+    
+    table_env.wait_exection_results()
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_run_main.py:84]:
```suggestion
    table_env.wait_execution_results()
```

**Review by SteNicholas — APPROVED** [review]

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_run_main.py (hunk 9 above)
Comment (lisy09) [reply flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_run_main.py]: fix with https://github.com/alibaba/flink-ai-extended/pull/477/commits/e7eab17b086ec11010b76fc322c510784b95b40d

**Review by SteNicholas — APPROVED** [review]

**Comment by lisy09** [comment]
@SteNicholas @wuchaochen 
Based on some discussion between me and @wuchaochen , I add this commit https://github.com/alibaba/flink-ai-extended/pull/477/commits/3c7ede1d733cb63d3aa71f31c289f30724454e00 to modify the return type of `FlinkEnv.create_env` to `(ExecutionEnvironment, WrappedTableEnvironment, WrappedStatementSet)`.

It introduces the API change for FlinkEnv.
Please help to review the latest version.

**Comment by SteNicholas** [comment]
> Hi @SteNicholas @wuchaochen What about the review process about this PR?

@lisy09 , please update the corresponding caller in the wide and deep example.

**Comment by lisy09** [comment]
> > Hi @SteNicholas @wuchaochen What about the review process about this PR?
> 
> @lisy09 , please update the corresponding caller in the wide and deep example.

Just fix with this commit: https://github.com/alibaba/flink-ai-extended/pull/477/commits/a7488312eaafb3e3125a3a78e8592e830ef0c199

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    def wait_execution_results(self):
+        for r in self.execute_sql_results:
+            job_client = r.get_job_client()
+            if job_client is not None:
+                job_client.get_job_execution_result().result()
+
+
+class WrappedTableEnvironment(TableEnvironment):
+    """
+    WrappedTableEnvironment override these functions to collect information need for flink_run_main:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:55]:
```suggestion
    WrappedTableEnvironment overrides the following functions to collect information needed for 'flink_run_main.py':
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    def create_from(t_env: StreamTableEnvironment) -> 'WrappedStreamTableEnvironment':
+        return WrappedStreamTableEnvironment(t_env._j_tenv)
+
+
+class WrappedStatementSetContext:
+    """
+    WrappedStatementSetContext provides container to store extra information for WrappedStatementSet,
+    including list of TableResult for each execution.
+    It can help to provide these functions:
+    - list of flink job id
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:99]:
```suggestion
    - Lists the ids of the Flink jobs.
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    - wait for execution result
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:100]:
```suggestion
    - Waits for the execution result of the Flink job.
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    - whether StatementSet needs to call execute() method based on whether it cantains not executed sql
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:101]:
```suggestion
    - Decides whether 'StatementSet' needs to call execute()' method based on whether it contains 'execute_sql'.
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    def wait_execution_results(self):
+        for r in self.execute_results:
+            job_client = r.get_job_client()
+            if job_client is not None:
+                job_client.get_job_execution_result().result()
+
+
+class WrappedStatementSet(StatementSet):
+    """
+    WrappedStatementSet override these functions to collect information need for flink_run_main:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_wrapped_env.py:128]:
```suggestion
    WrappedStatementSet overrides the following functions to collect information needed for 'flink_run_main.py':
```

File: flink-ai-flow/ai_flow_plugins/tests/job_plugins/ut_workflows/workflows/test_flink/test_flink.py [hunk 15]
```diff
@@ -77,6 +77,46 @@ def test_stop_local_flink_task(self):
         self.assertEqual(Status.FAILED, je.status)
         self.assertTrue('err' in je.properties)
 
+    def test_local_flink_no_statementset_task(self):
+        """
+        This is a test case for issue 475.
+        https://github.com/alibaba/flink-ai-extended/issues/475
+        It calls TableEnv.execute_sql in FlinkPythonProcessor and no need to call
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/tests/job_plugins/ut_workflows/workflows/test_flink/test_flink.py:84]:
```suggestion
        It calls 'TableEnv.execute_sql' in FlinkPythonProcessor and no need to call 
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]

**Review by SteNicholas — APPROVED** [review]

**Review by wuchaochen — APPROVED** [review]

**Pull request merged**